File: algorithms/2685. Count the Number of Complete Components.py
class Graph:
    def __init__(self, n: int, edges: List[List[int]]):
        self.n = n
        self.adj = [[] for _ in range(n)]
        self.ind = [0] * n
        for u, v in edges:
            self.adj[u].append(v)
            self.adj[v].append(u)
            self.ind[u] += 1
            self.ind[v] += 1

    def dfs(self, u, vis, comp):
        vis[u] = True
        comp.append(u)
        for v in self.adj[u]:
            if not vis[v]:
                self.dfs(v, vis, comp)

    def getConnectedComponents(self):
        components = []
        vis = [False] * self.n
        for u in range(self.n):
            if not vis[u]:
                comp = []
                self.dfs(u, vis, comp)
                components.append(comp)
        return components


def countCompleteComponents(n: int, edges: List[List[int]]) -> int:
    g = Graph(n, edges)
    components = g.getConnectedComponents()
    completeCompCount = 0
    for comp in components:
        numOfVertices = len(comp)
        if all(g.ind[x] == numOfVertices - 1 for x in comp):
            completeCompCount += 1
    return completeCompCount

File: algorithms/cpp/2685. Count the Number of Complete Components.cpp
#include <vector>
#include <set>
using namespace std;

class Graph {
public:
    int n;
    vector<vector<int>> adj;
    vector<int> ind;

    Graph(int n, const vector<vector<int>>& edges) : n(n), adj(n), ind(n, 0) {
        for (const auto& e : edges) {
            int u = e, v = e[1];
            adj[u].push_back(v);
            adj[v].push_back(u);
            ind[u]++;
            ind[v]++;
        }
    }

    void dfs(int u, vector<bool>& vis, vector<int>& comp) {
        vis[u] = true;
        comp.push_back(u);
        for (int v : adj[u])
            if (!vis[v]) dfs(v, vis, comp);
    }

    vector<vector<int>> getConnectedComponents() {
        vector<vector<int>> components;
        vector<bool> vis(n, false);
        for (int u = 0; u < n; ++u) {
            if (!vis[u]) {
                vector<int> comp;
                dfs(u, vis, comp);
                components.push_back(comp);
            }
        }
        return components;
    }
};

int countCompleteComponents(int n, vector<vector<int>>& edges) {
    Graph g(n, edges);
    auto components = g.getConnectedComponents();
    int completeCompCount = 0;
    for (auto& comp : components) {
        int numOfVertices = comp.size();
        set<int> indVals;
        for (int x : comp) indVals.insert(g.ind[x]);
        if (indVals.size() == 1 && *indVals.begin() == numOfVertices - 1) {
            completeCompCount++;
        }
    }
    return completeCompCount;
}
